_posts/2022-10-01-arrow-parquet-encoding-part-1.md

```diff
@@ -0,0 +1,152 @@
+author: tustvold, alamb
+categories: [parquet, arrow]
+---
+<!--
+{% comment %}
+Licensed to the Apache Software Foundation (ASF) under one or more
+contributor license agreements.  See the NOTICE file distributed with
+this work for additional information regarding copyright ownership.
+The ASF licenses this file to you under the Apache License, Version 2.0
+(the "License"); you may not use this file except in compliance with
+the License.  You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+{% endcomment %}
+-->
+
+## Introduction
+
+We recently completed a long-running project within [Rust Apache Arrow](https://github.com/apache/arrow-rs) to complete support for reading and writing arbitrarily nested Parquet and Arrow schemas. This is a complex topic, and we encountered a lack of approachable technical information, and thus wrote this blog to share our learnings with the community.
+
+[Apache Arrow](https://arrow.apache.org/) is an open, language-independent columnar memory format for flat and hierarchical data, organized for efficient analytic operations. [Apache Parquet](https://parquet.apache.org/) is an open, column-oriented data file format designed for very efficient data encoding and retrieval.
+
+It is increasingly common for analytic systems to use Arrow to process data stored in Parquet files, and therefore fast, efficient, and correct translation between them is a key building block.
+
+Historically analytic processing primarily focused on querying data with a tabular schema, where there are a fixed number of columns, and each row contains a single value for a each column. However, with the increasing adoption of structured document formats such as XML, JSON, etc…, this schema limitation can seem antiquated and unnecessarily limiting.
+
+As of version [20.0.0](https://crates.io/crates/arrow/20.0.0), released in August 2022, the Rust Arrow implementation for reading structured types is feature complete. Instructions for getting started can be found [here](https://docs.rs/parquet/latest/parquet/arrow/index.html) and feel free to raise any issues on our [bugtracker](https://github.com/apache/arrow-rs/issues).
+
+In this series we will explain how Parquet and Arrow represent nested data, highlighting the similarities and differences between them, and giving a flavor of the practicalities of converting between the formats.
+
+## Columnar vs Record-Oriented
+
+First, it is necessary to take a step back and discuss the difference between columnar and record-oriented data formats. In a record oriented data format, such as JSON, all the values for a given record are stored contiguously.
```

Comment:
I don't think JSON or other formats are inherently record-oriented or column-oriented. Also, depending on how pedantic you want to be, "column-oriented" or "record-oriented" are probably better described as data layouts; not "formats". I believe other sources describe them similarly (I just checked [this wiki page](https://en.wikipedia.org/wiki/Column-oriented_DBMS)).

I think it's better to change the description to say "layout," which may be defined along the lines of "how the physical bytes are stored in memory or on storage." You can keep these examples as is, but maybe clarify in the direction of, "When using a record-oriented layout, values for a given record are stored contiguously. Here's how we might represent this data using the JSON format."

Ultimately, I understand being terse, but it should be clear to the reader that a column-oriented layout can be represented in JSON because it is a format with no constraints on columns or records:
```json
{
    "Column1": [1, 3, 5],
    "Column2": [2, 4, 4],
    "Column3": [null, 5, 5]
}
```

Reply:
(also, a few paragraphs down "layout" is used, so yeah)

Reply:
```suggestion
First, it is necessary to take a step back and discuss the difference between columnar and record-oriented data formats. In a record oriented data format, such as newline-delimited JSON (NDJSON), all the values for a given record are stored contiguously.
```

I was specifically referring to ndjson here. Does that allay your concerns?

Reply:
njson is a record-oriented format, parquet is a column-oriented format, is this incorrect?

Reply:
I see. I didn't know about NDJSON specifically assuming records. Its format still doesn't preclude a column-oriented representation, but its assumption does.

then, you're right and this can be ignored.